## Loading data

In [1]:
english = {}
with open('./europarl_raw/test.en') as f:
    for line in f:
        for word in line.split():
            if word in english.keys():
                english[word] += 1
            else:
                english[word] = 1

In [2]:
german = {}
with open('./europarl_raw/test.de') as f:
    for line in f:
        for word in line.split():
            if word in german.keys():
                german[word] += 1
            else:
                german[word] = 1

## Counting no. words

In [3]:
print('No. English words: ', sum(english.values()))
print('No. German words: ', sum(german.values()))

print('No. English word types: ', len(english.keys()))
print('No. German word types: ', len(german.keys()))

No. English words:  6293
No. German words:  5752
No. English word types:  1572
No. German word types:  1798


## Vocab sizes

In [4]:
english_unks = [key for key, value in english.items() if value==1]
german_unks = [key for key, value in german.items() if value==1]
print('No. English <UNK>s: ', len(english_unks))
print('No. German <UNK>s: ', len(german_unks))

print('English vocab size: ', len(english.keys()) - len(english_unks) + 1)
print('German vocab size: ', len(german.keys()) - len(german_unks) + 1)

print('Total vocab size: ', len(english.keys()) - len(english_unks) + len(german.keys()) - len(german_unks) + 1)

No. English <UNK>s:  1016
No. German <UNK>s:  1290
English vocab size:  557
German vocab size:  509
Total vocab size:  1065


## Unknown words

In [5]:
english_unks

['breach',
 'manner',
 'misinterpretation',
 'abuse',
 'simplifying',
 'used',
 'lannoye',
 'moving',
 'adjourned',
 'reach',
 'voice',
 'her',
 'supporters',
 'firmly',
 'trends',
 'service',
 'rome',
 'proof',
 'success',
 'dangerous',
 'among',
 'romania',
 'instance',
 'criticism',
 'timeless',
 'speech',
 'binding',
 'international',
 'proved',
 'relating',
 'randzio',
 'plath',
 'exclusive',
 'ignores',
 'applying',
 'agents',
 'farm',
 'sure',
 'creation',
 'ethical',
 'space',
 'permitted',
 'ceiling',
 '1.27',
 'combined',
 'gnp',
 'poor',
 'political',
 'plot',
 'subject',
 'performance',
 'automobile',
 'industry',
 'regards',
 'co2',
 'emissions',
 'probability',
 'malta',
 'fossil',
 'reserves',
 'beneath',
 'seabed',
 'produced',
 'win',
 'failing',
 'worked',
 'farmers',
 'intended',
 'sections',
 'placed',
 'identify',
 'go',
 'beyond',
 'midnight',
 'base',
 'wife',
 'blind',
 'activity',
 'outside',
 'marital',
 'seek',
 'expert',
 'reports',
 'precondition',
 'likewi

In [6]:
german_unks

['abweichungen',
 'umgehungen',
 'tür',
 'tor',
 'öffnen',
 'arbeiten',
 'drastischen',
 'derzeitigen',
 'instrumentariums',
 'lannoye',
 'richtig',
 'beantragen',
 'vertagung',
 'wichtigkeit',
 'angelegenheit',
 'weiterer',
 'wichtiger',
 'angesprochen',
 'nahost-konflikt',
 'ihr',
 'anhängern',
 'festen',
 'überzeugung',
 'entwicklungen',
 'ergreifen',
 'öffentliche',
 'dienst',
 'eigentlichen',
 'römischen',
 'vertrags',
 'erwähnt',
 'fünfte',
 'jüngste',
 'erfolg',
 'beobachter',
 'gefährliche',
 'verrichten',
 'rumänien',
 'gesetze',
 'minderheiten',
 'einziger',
 'kritikpunkt',
 'ausführungen',
 'zeitlos',
 'keins',
 'existiert',
 'beschlüsse',
 'völkerrechtlich',
 'bindend',
 'gleichem',
 'maße',
 'eigenen',
 'bewährt',
 'soeben',
 'wortlaut',
 'randzio-plath',
 'hierbei',
 'ausschließliche',
 'ignoriert',
 'einsatz',
 'mittels',
 'feld',
 'schaffung',
 'ethikraums',
 'zulässige',
 'obergrenze',
 'eu-haushalt',
 '1,27',
 'bsp',
 'sollen',
 'menschenmögliche',
 'sicherheit',
 'le

There seems to be a number of proper nouns which only appear once. Many numbers also don't get repeated.
Examples: 'Kilo', '1.27'

## Shared words

In [38]:
shared_words = [word for word, count in english.items() if word in german.keys() and count > 1 and german[word] > 1]
print(len(shared_words))
print(shared_words)

29
[',', '.', '?', 'text', 'also', 'in', 'problem', 'eu', '-', 'union', '%', ':', 'will', 'an', '(', ')', 'am', "'", 'so', 'was', '9', 'initiative', 'situation', 'reform', '!', '2020', 'moment', 'war', 'internet']


These words can be given the same token in the vocabulary. This would reduce the vocabulary size by 29, giving us smaller representations of words and automatically translating words that mean the same thing.